### PR TITLE
Allow browsing contexts to resolve to cross-origin windows

### LINKED ISSUE #15180: WindowProxy for browsing contexts whose active document is in another script thread
At the moment, when we resolve a `WindowProxy` to a `Window`, we return the most recent window that was active in the current script thread. This gives the wrong results in the case of a browsing context whose active document is in another script thread.

### PR DESCRIPTION
**Pull request by asajeffrey**
This PR implements cross-thread `WindowProxy` objects.

At the moment, if a `Window` performs a non-similar-origin navigation, the old script thread does not update its `WindowProxy`, since the new `Window` is in the new script thread. With this PR, the `WindowProxy` is updated to a dummy `XOriginWindow` object, that only implements the whitelisted methods that are allowed to be called cross-origin.

This PR does not include working implementations of some of the cross-origin `Window` or `Location` methods.

This PR causes some cross-origin wpt tests to now pass, in particular `/html/browsers/origin/cross-origin-objects/cross-origin-objects.html ` now passes `Only whitelisted properties are accessible cross-origin`. There are some CORS failures in `fetch`, I suspect caused by the incorrect setting of `origin` in fetch requests.

Although there are some functions that now throw `SecurityException`, it is not meant to be a complete implementation, which will have to wait for XOWs to land.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #15180.
- [X] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15358)

### CODE REVIEW

**Comment by asajeffrey** [comment]
cc @avadacatavra @bzbarsky @Ms2ger
r? @jdm

File: components/script/dom/bindings/reflector.rs [hunk 1]
```diff
@@ -50,6 +51,18 @@ impl Reflector {
...
+    #[inline]
+    pub fn get_runtime(&self) -> *mut JSRuntime {
+        unsafe { JS_GetObjectRuntime(self.object.get()) }
+    }
+
+    /// Get the JS context.
+    #[inline]
+    pub fn get_cx(&self) -> *mut JSContext {
+        unsafe { JS_GetContext(self.get_runtime()) }
+    }
```
Comment (Ms2ger) [inline components/script/dom/bindings/reflector.rs]: Why? We might as well use `Runtime::new()`; that's probably faster than two out-of-line calls.

File: components/script/dom/browsingcontext.rs [hunk 2]
```diff
@@ -61,7 +70,7 @@ impl BrowsingContext {
             let WindowProxyHandler(handler) = window.windowproxy_handler();
             assert!(!handler.is_null());
 
-            let cx = window.get_cx();
+            let cx = window.reflector().get_cx();
```
Comment (Ms2ger) [inline components/script/dom/browsingcontext.rs]: Why?

File: components/script/dom/browsingcontext.rs [hunk 3]
```diff
@@ -104,13 +114,13 @@ impl BrowsingContext {
     /// Change the Window that this browsing context's WindowProxy resolves to.
     // TODO: support setting the window proxy to a dummy value,
     // to handle the case when the active document is in another script thread.
-    pub fn set_window_proxy(&self, window: &Window) {
+    fn set_window_proxy<W: DomObject>(&self, window: &W, traps: &ProxyTraps) {
```
Comment (Ms2ger) [inline components/script/dom/browsingcontext.rs]: Allowing any `DomObject` here does not sound like a great strategy.

File: components/script/dom/browsingcontext.rs [hunk 4]
```diff
@@ -144,6 +154,21 @@ impl BrowsingContext {
         }
     }
 
+    pub fn set_currently_active(&self, window: &Window) {
+        self.set_window_proxy(window, &PROXY_HANDLER);
+        self.currently_active.set(Some(window.global().pipeline_id()));
+    }
+
+    pub fn unset_currently_active(&self) {
+        let win = XOriginWindow::new(self);
```
Comment (Ms2ger) [inline components/script/dom/browsingcontext.rs]: Would `CrossOriginWindow` be that much harder to type?

File: components/script/dom/xoriginlocation.rs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-location-reload
+    fn Reload(&self) -> Fallible<()> {
+        Err(Error::Security)
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-location-href
+    fn Stringifier(&self) -> DOMString {
+        DOMString::from("<XOriginWindow>")
```
Comment (Ms2ger) [inline components/script/dom/xoriginlocation.rs]: Uh...

File: components/script/dom/xoriginwindow.rs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+use dom::bindings::str::DOMString;
+use dom::browsingcontext::BrowsingContext;
+use dom::globalscope::GlobalScope;
+use dom::xoriginlocation::XOriginLocation;
+use ipc_channel::ipc;
+use js::jsapi::{JSContext, HandleValue};
+use js::jsval::{JSVal, UndefinedValue};
+use msg::constellation_msg::PipelineId;
+
+// TODO: documentation!
```
Comment (Ms2ger) [inline components/script/dom/xoriginwindow.rs]: Indeed.

File: tests/wpt/web-platform-tests/common/get-host-info.sub.js [hunk 7]
```diff
@@ -4,7 +4,7 @@ function get_host_info() {
   var HTTP_PORT2 = '{{ports[http][1]}}';
   var HTTPS_PORT = '{{ports[https][0]}}';
   var ORIGINAL_HOST = '{{host}}';
-  var REMOTE_HOST = (ORIGINAL_HOST === 'localhost') ? '127.0.0.1' : ('www1.' + ORIGINAL_HOST);
+  var REMOTE_HOST = (ORIGINAL_HOST === '127.0.0.1') ? '127.0.0.2' : '127.0.0.1';
```
Comment (Ms2ger) [inline tests/wpt/web-platform-tests/common/get-host-info.sub.js]: r? @jgraham

File: components/script/dom/bindings/reflector.rs (hunk 1 above)
Comment (asajeffrey) [inline components/script/dom/bindings/reflector.rs]: This probably isn't needed any more, originally I was fighting to try to get Servo to have globals that aren't `GlobalScope`s but eventually gave up. This looks like code that was left over from that fight.

File: components/script/dom/browsingcontext.rs (hunk 2 above)
Comment (asajeffrey) [inline components/script/dom/browsingcontext.rs]: Ditto.

File: components/script/dom/browsingcontext.rs (hunk 3 above)
Comment (asajeffrey) [inline components/script/dom/browsingcontext.rs]: Ditto. The problem was that when I started this, `XOriginWindow` was a global, but not a `GlobalScope`. Now it is, so I can make this a `GlobalScope`, so we can statically check that it's only called with globals.

File: components/script/dom/browsingcontext.rs (hunk 4 above)
Comment (asajeffrey) [inline components/script/dom/browsingcontext.rs]: Ooh, bikeshedding! `XThreadWindow`? `CrossThreadWindow`? `DissimillarOriginWindow`?

File: components/script/dom/xoriginlocation.rs (hunk 5 above)
Comment (asajeffrey) [inline components/script/dom/xoriginlocation.rs]: Er yes, not sure what to put here. Stringifiers aren't allowed to throw, so we have to return something. Once we have XOWs, this will never be called so we can `panic!`, but in the mean time we have to return some `DOMString`.

File: components/script/dom/xoriginwindow.rs (hunk 6 above)
Comment (asajeffrey) [inline components/script/dom/xoriginwindow.rs]: Oh yes, must write some of that.

File: tests/wpt/web-platform-tests/common/get-host-info.sub.js (hunk 7 above)
Comment (asajeffrey) [inline tests/wpt/web-platform-tests/common/get-host-info.sub.js]: @jgraham the reason for this change is to run the test suite with a non-similar `REMOTE_HOST`. At the moment, `ORIGINAL_HOST` and `REMOTE_HOST` are similar-origin, so we never actually run any tests cross-thread.

File: components/script/dom/xoriginlocation.rs (hunk 5 above)
Comment (bzbarsky) [inline components/script/dom/xoriginlocation.rs]: Er... Why are stringifiers not allowed to throw?  There's nothing that prevents a spec from defining a stringifier that sometimes throws.

Comment (asajeffrey) [inline components/script/dom/xoriginlocation.rs]: @bzbarsky D'oh! Fixed this, the stringifier now throws a security exception.

**Comment by jdm** [comment]
Looks like a `./mach update-manifest` is required.

File: components/script/dom/browsingcontext.rs [hunk 8]
```diff
@@ -144,6 +154,23 @@ impl BrowsingContext {
...
+    pub fn set_currently_active(&self, window: &Window) {
+        let globalscope = window.global();
+        self.set_window_proxy(&*globalscope, &PROXY_HANDLER);
+        self.currently_active.set(Some(globalscope.pipeline_id()));
+    }
+
+    pub fn unset_currently_active(&self) {
+        let window = XOriginWindow::new(self);
+        let globalscope = window.global();
+        self.set_window_proxy(&*globalscope, &XORIGIN_PROXY_HANDLER);
```
Comment (jdm) [inline components/script/dom/browsingcontext.rs]: This looks like a confusing way of writing:
```rust
let window = XOriginWindow::new(self);
self.set_window_proxy(&*window.upcast(), &XORIGIN_PROXY_HANDLER);
```

File: components/script/dom/browsingcontext.rs [hunk 9]
```diff
@@ -144,6 +154,23 @@ impl BrowsingContext {
         }
     }
 
+    pub fn set_currently_active(&self, window: &Window) {
+        let globalscope = window.global();
+        self.set_window_proxy(&*globalscope, &PROXY_HANDLER);
```
Comment (jdm) [inline components/script/dom/browsingcontext.rs]: `let globalscope = window.upcast();`

File: components/script/dom/browsingcontext.rs [hunk 10]
```diff
@@ -104,10 +114,10 @@ impl BrowsingContext {
     /// Change the Window that this browsing context's WindowProxy resolves to.
     // TODO: support setting the window proxy to a dummy value,
     // to handle the case when the active document is in another script thread.
-    pub fn set_window_proxy(&self, window: &Window) {
+    fn set_window_proxy(&self, window: &GlobalScope, traps: &ProxyTraps) {
         unsafe {
             debug!("Setting window proxy of {:p}.", self);
-            let WindowProxyHandler(handler) = window.windowproxy_handler();
+            let handler = CreateWrapperProxyHandler(traps);
```
Comment (jdm) [inline components/script/dom/browsingcontext.rs]: We shouldn't need to create a new handler every time here. We can create it once and store it in DomStaticData like the [regular](https://dxr.mozilla.org/servo/rev/287b02e21fa2c81d58b070be36add5e951512679/components/script/dom/bindings/utils.rs#61) windowproxy handler.

File: components/script/dom/xoriginlocation.rs [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+impl XOriginLocation {
+    #[allow(unsafe_code)]
+    pub fn new(window: &XOriginWindow) -> Root<XOriginLocation> {
+        let globalscope = window.global();
+        let cx = globalscope.get_cx();
+        let loc = box XOriginLocation {
+            reflector: Reflector::new(),
+            window: JS::from_ref(window),
+        };
+        unsafe {
```
Comment (jdm) [inline components/script/dom/xoriginlocation.rs]: I think we can use reflect_dom_object here and remove the unsafe_code allow.

File: components/script/dom/xoriginwindow.rs [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+impl XOriginWindow {
+    #[allow(unsafe_code)]
+    pub fn new(browsing_context: &BrowsingContext) -> Root<XOriginWindow> {
+        let globalscope = browsing_context.global();
+        let cx = globalscope.get_cx();
+        // Any timer events fired on this window are ignored.
+        let (timer_event_chan, _) = ipc::channel().unwrap();
+        let win = box XOriginWindow {
+            globalscope: GlobalScope::new_inherited(PipelineId::new(),
```
Comment (jdm) [inline components/script/dom/xoriginwindow.rs]: We may want a TODO here about storing the actual pipeline that this cross-origin window represents.

File: tests/wpt/web-platform-tests/common/get-host-info.sub.js (hunk 7 above)
Comment (jdm) [inline tests/wpt/web-platform-tests/common/get-host-info.sub.js]: Doesn't this break tests? REMOTE_HOST is supposed to be a URL that is actually reachable; I would expect the case that uses 127.0.0.2 to time out whenever it attempts to connect.

**Review by jdm — CHANGES_REQUESTED** [review]
Nothing particularly concerning here. Well done!

File: components/script/dom/webidls/XOriginWindow.webidl [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+// accessors that do not throw security exceptions when called
+// cross-origin.
+//
+// Note that similar-origin windows are kept in the same script
+// thread, so this mechanism cannot be relied upon as the only
+// way to enforce security policy.
+
+// https://html.spec.whatwg.org/multipage/#window
+[Global, NoInterfaceObject]
+interface XOriginWindow : GlobalScope {
```
Comment (asajeffrey) [inline components/script/dom/webidls/XOriginWindow.webidl]: Oh good point. Should I just file an issue for this?

One possible solution, but it would mean rewriting codegen, would be to allow more than one Servo implementation of a webidl class, in particular to make `Foo::Wrap` generic in any type which implements `FooMethods`, not specific to `Foo`. This isn't an easy change though.

File: components/script/dom/browsingcontext.rs (hunk 10 above)
Comment (asajeffrey) [inline components/script/dom/browsingcontext.rs]: Ah, is `CreateWrapperProxyHandler` expensive?

File: components/script/dom/browsingcontext.rs (hunk 9 above)
Comment (asajeffrey) [inline components/script/dom/browsingcontext.rs]: Fixed.

File: components/script/dom/browsingcontext.rs (hunk 8 above)
Comment (asajeffrey) [inline components/script/dom/browsingcontext.rs]: Fixed.

File: components/script/dom/xoriginlocation.rs (hunk 11 above)
Comment (asajeffrey) [inline components/script/dom/xoriginlocation.rs]: Fixed. This was a holdover from `XOriginWindow` not implementing `GlobalScope`.

File: components/script/dom/xoriginwindow.rs (hunk 12 above)
Comment (asajeffrey) [inline components/script/dom/xoriginwindow.rs]: We could do that for the initial pipeline id, but to keep that id up-to-date, we'd have to send a lot of messages around (N messages ever time the pipeline changed) which seems not worth it. I'll add a comment about this.

**Comment by asajeffrey** [comment]
For some reason, github isn't letting me reply to https://github.com/servo/servo/pull/15358#discussion_r100193847

Good point, there's no http server running on 127.0.0.2. As a patch, I can set it to `web-platform.test`, but it would be nice if we had two dissimilar-origin host names. @jgraham?

File: components/script/dom/browsingcontext.rs (hunk 10 above)
Comment (jdm) [inline components/script/dom/browsingcontext.rs]: It's more that we only need a singleton, and creating more will lead to memory leaks if we don't clean them up.

File: components/script/dom/webidls/XOriginWindow.webidl (hunk 13 above)
Comment (jdm) [inline components/script/dom/webidls/XOriginWindow.webidl]: An issue would be good.

**Comment by jdm** [comment]
Yeah, http://testthewebforward.org/docs/test-format-guidelines.html#tests-involving-multiple-origins only describes subdomains unfortunately.

**Comment by jgraham** [comment]
Well, https and multiple ports are supposed to give you dissimilar origins. I still don't really understand why servo can't use those things as a part of the iframe process heuristics? But if we want another top level host, that should be possible but will require more coordination to check that we have buy-in.

**Comment by bzbarsky** [comment]
> Well, https and multiple ports are supposed to give you dissimilar origins.

Multiple ports is not enough to give you dissimilar origins.  If you have "http://foo.com:10" and "http://foo.com:20" and they both set document.domain to "foo.com", they end up same origin-domain.

https is enough, though.

File: components/script/dom/webidls/XOriginWindow.webidl (hunk 13 above)
Comment (asajeffrey) [inline components/script/dom/webidls/XOriginWindow.webidl]: Filed #15476.

File: components/script/dom/browsingcontext.rs (hunk 10 above)
Comment (asajeffrey) [inline components/script/dom/browsingcontext.rs]: Won't they just get gc'd when the window proxy is unreachable?

**Comment by asajeffrey** [comment]
I don't think we want to require all our cross-origin tests to run with one server http and one https.

File: components/script/dom/browsingcontext.rs [hunk 14]
```diff
@@ -45,6 +46,9 @@ pub struct BrowsingContext {
 
     /// The pipeline id of the currently active document.
     /// May be None, when the currently active document is in another script thread.
+    /// We do not try to keep the pipeline id for documents in other threads,
+    /// as this would require the constelltion notifying many script threads about
```
Comment (jdm) [inline components/script/dom/browsingcontext.rs]: s/constelltion/constellation/

File: components/script/dom/xoriginlocation.rs [hunk 15]
```diff
@@ -29,18 +29,19 @@ pub struct XOriginLocation {
 }
 
 impl XOriginLocation {
-    #[allow(unsafe_code)]
-    pub fn new(window: &XOriginWindow) -> Root<XOriginLocation> {
-        let globalscope = window.global();
-        let cx = globalscope.get_cx();
-        let loc = box XOriginLocation {
+    #[allow(unrooted_must_root)]
+    pub fn new_inherited(window: &XOriginWindow) -> XOriginLocation {
```
Comment (jdm) [inline components/script/dom/xoriginlocation.rs]: This should not be public.

File: tests/wpt/mozilla/meta/MANIFEST.json [hunk 16]
```diff
@@ -25940,6 +25945,10 @@
    "333a80e3d4aec4a8be5566605fae5675a821085b",
    "support"
   ],
+  "mozilla/referrer-policy/generic/subresource/mozresource.pyc": [
+   "c3d7573593042e1d218e532071225d92c4920954",
+   "support"
+  ],
```
Comment (jdm) [inline tests/wpt/mozilla/meta/MANIFEST.json]: You'll need to remove the pyc files from your working tree and run manifest-update again.

**Review by jdm — CHANGES_REQUESTED** [review]
@jgraham will need to sign off on the get-host-info.sub.js change.

File: components/script/dom/browsingcontext.rs (hunk 14 above)
Comment (asajeffrey) [inline components/script/dom/browsingcontext.rs]: Done.

File: components/script/dom/xoriginlocation.rs (hunk 15 above)
Comment (asajeffrey) [inline components/script/dom/xoriginlocation.rs]: Done.

File: tests/wpt/mozilla/meta/MANIFEST.json (hunk 16 above)
Comment (asajeffrey) [inline tests/wpt/mozilla/meta/MANIFEST.json]: Done.

**Comment by asajeffrey** [comment]
OK, @jgraham the one remaining issue is how to run dissimilar-origin tests.

**Review by jdm — APPROVED** [review]

**Comment by jdm** [comment]
This looks good to merge after appropriate squashing.

**Comment by asajeffrey** [comment]
I'm going to add a copy of the cross-origin-objects test that uses dissimilar-origin URLs to the Mozilla wpt directory, then do the landing dance.

**Comment by asajeffrey** [comment]
IRC conversation about testing with @jgraham and @jdm: http://logs.glob.uno/?c=mozilla%23servo&s=13+Feb+2017&e=13+Feb+2017#c610983

TL;DR: in the long term, add another host to wpt so that tests can run with dissimilar-origin URLs. In the short term, add a servo-specific dissimilar-origin version of `cross-origin-objects.html`.

**Comment by asajeffrey** [comment]
Squashed. @bors-servo r=jdm

**Comment by asajeffrey** [comment]
Hmm, bors seems not to be listening to me. Updated test expectations.

@bors-servo r=jdm

**Comment by asajeffrey** [comment]
http://build.servo.org/builders/linux-rel-wpt/builds/2391/steps/shell__1/logs/filtered-wpt-errorsummary.log:
```
{"status": "CRASH", "stack": null, "subtest": null, "action": "test_result", "test": "/navigation-timing/nav2_test_open_data_uri.html", "line": 202462, "message": null, "expected": "OK"}
{"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/_mozilla/mozilla/cross-origin-objects/cross-origin-objects.html", "line": 244276, "message": null, "expected": "OK"}
```
which are:
```
  ▶ CRASH [expected OK] /navigation-timing/nav2_test_open_data_uri.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ ERROR:script::dom::bindings::error: Error at data:text/html;charset=utf-8,%3C%21DOCTYPE%20html%3E%0D%0A%3Ctitle%3Edata%20URL%20source%20for%20navigation-timing%2Fnav2_test_open_data_uri.html%3C%2Ftitle%3E%0D%0A%3C%21--%20NB%3A%20this%20file%20isn%27t%20actually%20used%20any%20where%21%20--%3E%0D%0A%3Clink%20rel%3D%22author%22%20title%3D%22Google%22%20href%3D%22http%3A%2F%2Fwww.google.com%2F%22%20%2F%3E%0D%0A%3Cscript%3E%0D%0Avar%20observer%20%3D%20new%20PerformanceObserver%28%0D%0A%20%20%20%20function%20%28entryList%29%20%7B%0D%0A%20%20%20%20%20%20%20%20parent.postMessage%28%22observed%22%2C%20%22%2A%22%29%3B%0D%0A%20%20%20%20%20%20%20%20observer.disconnect%28%29%3B%0D%0A%20%20%20%20%7D%29%3B%0D%0Aobserver.observe%28%7BentryTypes%3A%20%5B%22navigation%22%5D%7D%29%3B%0D%0A%3C%2Fscript%3E%0D%0A:6:5 PerformanceObserver is not defined
  │ called `Result::unwrap()` on an `Err` value: Error { repr: Custom(Custom { kind: UnexpectedEof, error: StringError("failed to fill whole buffer") }) } (thread RenderBackend, at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libcore/result.rs:868)
  │ stack backtrace:
  │    0:     0x7f925492140d - backtrace::backtrace::trace::h976012eed768c7ff
  │    1:     0x7f92549218f2 - backtrace::capture::Backtrace::new::hb5a725a088a2a2fc
  │    2:     0x7f92532933c7 - servo::main::{{closure}}::h039ba3c803113243
  │    3:     0x7f925510c318 - std::panicking::rust_panic_with_hook
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:556
  │    4:     0x7f925510c1e4 - std::panicking::begin_panic<collections::string::String>
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:517
  │    5:     0x7f925510c109 - std::panicking::begin_panic_fmt
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:501
  │    6:     0x7f925510c097 - std::panicking::rust_begin_panic
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:477
  │    7:     0x7f925513985d - core::panicking::panic_fmt
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libcore/panicking.rs:69
  │    8:     0x7f925496a05d - core::result::unwrap_failed::h4e1dafabda202b02
  │    9:     0x7f92549bf12e - webrender::render_backend::RenderBackend::run::h806d46e08283bc83
  │   10:     0x7f925495ba6c - std::panicking::try::do_call::h7d57ef86fec8f13f
  │   11:     0x7f925511325a - panic_unwind::__rust_maybe_catch_panic
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libpanic_unwind/lib.rs:98
  │   12:     0x7f9254973657 - <F as alloc::boxed::FnBox<A>>::call_box::h232db978e3695021
  │   13:     0x7f925510afa4 - alloc::boxed::{{impl}}::call_once<(),()>
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/liballoc/boxed.rs:623
  │                          - std::sys_common::thread::start_thread
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/sys_common/thread.rs:21
  │                          - std::sys::imp::thread::{{impl}}::new::thread_start
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/sys/unix/thread.rs:84
  │   14:     0x7f9251160183 - start_thread
  │   15:     0x7f9250c7737c - clone
  │   16:                0x0 - <unknown>
  │ ERROR:servo: called `Result::unwrap()` on an `Err` value: Error { repr: Custom(Custom { kind: UnexpectedEof, error: StringError("failed to fill whole buffer") }) }
  └ Pipeline failed in hard-fail mode.  Crashing!

  ▶ TIMEOUT [expected OK] /_mozilla/mozilla/cross-origin-objects/cross-origin-objects.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1
```

**Comment by asajeffrey** [comment]
The first one of those is #15281.

**Comment by asajeffrey** [comment]
Can't recreate the `TIMEOUT` locally.

@bors-servo: retry
 * #15281

**Comment by asajeffrey** [comment]
Okay, my guess is that this test is genuinely timing out on the CI hardware, I'll try splitting it into subtests and see if that helps.

**Comment by asajeffrey** [comment]
I radically trimmed back `/_mozilla/mozilla/cross-origin-objects/cross-origin-objects.html`, hopefully it won't `TIMEOUT` on the CI hardware now.

@bors-servo r=jdm

**Comment by asajeffrey** [comment]
Hmm:
```
 ▶ Unexpected subtest result in /_mozilla/mozilla/cross-origin-objects/cross-origin-objects.html:
  └ PASS [expected FAIL] Parentage of cross-origin windows

  ▶ Unexpected subtest result in /_mozilla/mozilla/cross-origin-objects/cross-origin-objects.html:
  │ FAIL [expected PASS] Only whitelisted properties are accessible cross-origin
  │   → assert_throws: Should throw when writing to self on Window function "function () { C[prop] = undefined; }" did not throw
  │ 
  │ @http://web-platform.test:8000/_mozilla/mozilla/cross-origin-objects/cross-origin-objects.html:72:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1409:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1433:20

```
The first one is #14775, the second is caused by the test running while `C` is same-origin, when it should be dissimilar-origin. This may be a timing issue, caused by reacting to the `load` event from the initial `about:blank`. I'll try rewriting the test and running it again.

**Comment by asajeffrey** [comment]
OK, I rewrote the test to be more robust against timing variations. @bors-servo r=jdm
